odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2385 @@
+        >>> space = odl.cn((2, 3))
+        >>> x = space.element([[1 + 1j, 2, 3 - 3j],
+        ...                    [4, 5 - 5j, 6]])
+        >>> x.conj()
+        cn((2, 3)).element(
+            [[(1-1j), (2-0j), (3+3j)],
+             [(4-0j), (5+5j), (6-0j)]]
+        )
+
+        The out parameter allows you to avoid a copy
+
+        >>> y = space.element()
+        >>> y_out = x.conj(out=y)
+        >>> y
+        cn((2, 3)).element(
+            [[(1-1j), (2-0j), (3+3j)],
+             [(4-0j), (5+5j), (6-0j)]]
+        )
+        >>> y_out is y
+        True
+
+        It can also be used for in-place conjugation:
+
+        >>> x_out = x.conj(out=x)
+        >>> x
+        cn((2, 3)).element(
+            [[(1-1j), (2-0j), (3+3j)],
+             [(4-0j), (5+5j), (6-0j)]]
+        )
+        """
+        if out is None:
+            return self.space.element(self.data.conj())
+        else:
+            self.data.conj(out.data)
+            return out
+
+    def __getitem__(self, indices):
+        """Return ``self[indices]``.
+
+        Parameters
+        ----------
+        indices : index expression
+            Integer, slice or sequence of these, defining the positions
+            of the data array which should be accessed.
+
+        Returns
+        -------
+        values : `NumpyTensorSet.dtype` or `NumpyGeneralizedTensor`
+            The value(s) at the given indices. Note that the returned
+            object is a writable view into the original tensor.
+
+        Examples
+        --------
+        >>> space = odl.rn((2, 3))
+        >>> x = space.element([[1, 2, 3],
+        ...                    [4, 5, 6]])
+        >>> x[0, 1]
+        2.0
+        >>> x[:, 1:]
+        rn((2, 2)).element(
+            [[2.0, 3.0],
+             [5.0, 6.0]]
+        )
+        """
+        arr = self.data[indices]
+        if np.isscalar(arr):
+            return arr
+        else:
+            space_constructor = type(self.space)
+            if (self.order in ('C', 'F') and
+                    arr.flags[self.order + '_CONTIGUOUS']):
+                out_spc_order = self.order
+            else:
+                # To preserve the array view for non-contiguous slices,
+                # we need to use 'K' for the space in that case.
+                out_spc_order = 'K'
+            space = space_constructor(
+                arr.shape, dtype=self.dtype, order=out_spc_order,
+                exponent=self.space.exponent, weighting=self.space.weighting)
+            return space.element(arr)
+
+    def __ipow__(self, other):
+        """Return ``self **= other``."""
+        try:
+            if other == int(other):
+                return super().__ipow__(other)
+        except TypeError:
+            pass
+
+        np.power(self.data, other, out=self.data)
+        return self
+
+    def __int__(self):
+        """Return ``int(self)``.
+
+        Raises
+        ------
+        TypeError
+            if ``self.size != 1``.
```

Comment:
Actually many of these sections don't make sense since they basically document NumPy. What if they decide to change the exception type?